Plans data-store: refactor unit tests and fix TypeScript errors

ciampo opened the pull request:
## Changes proposed in this Pull Request

* Add mock data for tests that mimics real-world data
* Fix TypeScript errors
* Change a few types based on the data coming from API endpoints (including making `path_slug` optional). Types from APIs data have been simplified (only properties that are actually used are typed, in order to make it more explicit)

## Testing instructions

- `yarn test-packages:watch packages/data-stores/src/plans` — the plans data-store tests should pass
- Smoke test plans grid on Gutenboarding, Step by Step launch and Focused Launch
  - For good measure, open all plans grid ts/tsx files and check that there are not TypeScript errors

Closes #49536

ciampo committed:
Type action generators

ciampo committed:
Add DetailsAPIFeature type

ciampo committed:
Update plan data-store types (missing props, mismatching types...)

ciampo committed:
Add mock data in a separate file

ciampo committed:
Use mock data in tests

ciampo committed:
Refactor mock data in sections, add comment

ciampo committed:
Fix resolver types

ciampo commented on packages/data-stores/src/plans/mock/mock-constants.ts:
```diff
@@ -0,0 +1,477 @@
+};
+
+// Individual plan (from APIs)
+// @TODO: path_slug doesn't exist on monthly and free plans
```
While gathering mock data from the API endpoints, I realised that the data for the free and the monthly-billed plans doesn't contain the `path_slug` property.

This is causing a TypeScript error, since we expect this value to be always defined (see `PricedAPIPlan` and, consequently, `PlanProduct` types), and has potential consequence in Gutenboarding, where we expect `path_slug` to always be defined.

I see two potential ways of getting around this:
1. The cleanest way would be to update the REST APIs and include `path_slug` also in the Free and monthly-billed plans
2. As an alternative, we could update the `PricedAPIPlan` type and mark `path_slug` as optional. Then, in the data-store, when building `PlanProduct` objects, we can use the "agnostic" slug so that we can ensure that the `PlanProduct` type always has a defined `path_slug`

razvanpapadopol replied:
We're already doing the alternative (2) for _Free_ plan: https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/resolvers.ts#L154

I don't have a strong opinion on this but whatever solution you implement, you should also make sure the bugs fixed in these PRs are not coming back 😅 :
- https://github.com/Automattic/wp-calypso/pull/49392
- https://github.com/Automattic/wp-calypso/pull/49426

And as a bonus point, increasing unit test coverage to prevent future bugs like these would be awesome.

StefanNieuwenhuis replied:
Follow-up question: Do both plans need the `path_slug` prop? If not, then it might be a good idea to split up `PricedAPIPlan`, like we suggest doing for `DomainSuggestions` here #49540. 

I would definitely go for option 1 since we won't deviate from the API.

ciampo replied:
> We're already doing the alternative (2) for Free plan: https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/plans/resolvers.ts#L154

We _sort of_ are, but it's still hacky, because TypeScript still thinks that a Free plan returned from the APIs should have a `path_slug` property

> I don't have a strong opinion on this but whatever solution you implement, you should also make sure the bugs fixed in these PRs are not coming back

> Follow-up question: Do both plans need the path_slug prop? If not, then it might be a good idea to split up PricedAPIPlan, like we suggest doing for DomainSuggestions here #49540.

I'll explore the two alternatives and make a more informed decicion

> And as a bonus point, increasing unit test coverage to prevent future bugs like these would be awesome.

Absolutely. I'll look into adding more tests to this PR too

razvanpapadopol replied:
> Do both plans need the path_slug prop?

On a functional level, they kind of do. We have also https://wordpress.com/new/free flow alongside the flows for paid plans like https://wordpress.com/new/business

ciampo commented on packages/data-stores/src/plans/test/resolvers.ts:
```diff
@@ -14,69 +34,10 @@ describe( 'getSupportedPlans', () => {
-				},
-			],
-		};
+		// Prepare stricter iterator types
+		type IteratorReturnType = ReturnType< typeof iter.next >;
+		type PlanPriceApiDataIterator = ( planPriceData: PricedAPIPlan[] ) => IteratorReturnType;
+		type PlanDetailsApiDataIterator = ( { body: DetailsAPIResponse } ) => IteratorReturnType;
```
This is to "help" TypeScript to expect the correct arguments when invoking `iter.next()` later on in the tests

StefanNieuwenhuis reviewed (commented):
First of all, I love the effort you put into the tests and mocks but I'm a bit worried about the maintainability. Currently, the tests are coupled to the state object structure through `.toEqual()` assertions (e.g. [`expect( features[ PLAN_FREE ].name ).toBe( MOCK_SIMPLIFIED_FEATURE_PRIORITY_SUPPORT.name );`](https://github.com/Automattic/wp-calypso/pull/49595/files#diff-66a8493c9e9dd2ffbf4847813192602672e45ba612ef9c1762a46a6628cf0f63R89). 

When we have to update our reducer (e.g. the introduction of monthly plans) it also requires us to update multiple test cases too. Depending on the changes, it'll be a dirty job.

What if we use selectors to check the reducer's outcomes? In this way, the tests are more decoupled from the implementation. Because, in the end, what we want to test is if we are correctly dispatching actions and selecting the right data from the store(s), and with selectors, we're more than able to do so. 

That aside, in order to test our reducer in isolation (i.e. without testing the API/side-effects), we still have to mock the API responses. 

Here's [an article on this testing strategy](https://dev.to/hbarcelos/a-better-approach-for-testing-your-redux-code-2ec9). What do you think?

ciampo commented:
> First of all, I love the effort you put into the tests and mocks but I'm a bit worried about the maintainability. Currently, the tests are coupled to the state object structure through `.toEqual()` assertions (e.g. [`expect( features[ PLAN_FREE ].name ).toBe( MOCK_SIMPLIFIED_FEATURE_PRIORITY_SUPPORT.name );`](https://github.com/Automattic/wp-calypso/pull/49595/files#diff-66a8493c9e9dd2ffbf4847813192602672e45ba612ef9c1762a46a6628cf0f63R89).
> 
> When we have to update our reducer (e.g. the introduction of monthly plans) it also requires us to update multiple test cases too. Depending on the changes, it'll be a dirty job.
> 
> What if we use selectors to check the reducer's outcomes? In this way, the tests are more decoupled from the implementation. Because, in the end, what we want to test is if we are correctly dispatching actions and selecting the right data from the store(s), and with selectors, we're more than able to do so.
> 
> That aside, in order to test our reducer in isolation (i.e. without testing the API/side-effects), we still have to mock the API responses.
> 
> Here's [an article on this testing strategy](https://dev.to/hbarcelos/a-better-approach-for-testing-your-redux-code-2ec9). What do you think?

I tend to agree on most of your comments! This PR is only a refactor of the current tests, mainly aimed at removing all of the TypeScript errors that we had (which made it harder to understand and assess the correctness of the tests).

As I added mocked data from real API calls, I also realised that some of the Types are not 100% correct, and so I'll have to include those changes in this PR too.

But otherwise, it would be probably better to keep this PR to only be a "relatively" small refactor, and change/add more tests in a separate PR?

StefanNieuwenhuis commented:
> But otherwise, it would be probably better to keep this PR to only be a "relatively" small refactor, and change/add more tests in a separate PR?

I agree that we should keep PR's as small as possible and prevent scope creep wherever we can. Let's fix this in a new PR while adding new/changing  tests.

ciampo committed:
Allow path_slug not to be defined on all plans

ciampo commented:
This PR is ready for review.

---

We decided to split the work ahead in smaller steps:
1. open a phab patch to path_slug to monthly plans
2. discuss if we should add path_slug to the free plan as well
3. In the meantime, open a new PR to split Plan types into free vs paid plans (and potentially add path_slug changes from previous points)
4. When all of this is done, open more PRs to re-work tests, and add more coverage

ciampo commented:
Once approved, probably better to wait for #49759 to be merged first, to solve any potential conflicts

ciampo committed:
Simplify PricedAPIPlan, DetailsAPIResponse and APIPlanDetail types

- Only keep properties that are used in code
- Add more specific PricedAPIPlan types (free, paid annually, paid montly)

StefanNieuwenhuis commented on packages/data-stores/src/plans/actions.ts:
```diff
@@ -3,47 +3,64 @@
+	type: 'RESET_PLAN' as const,
+} );
 
 export type PlanAction = ReturnType<
```
[Nit pick]: Doesn't it make more sense to use a plural here since we exporting multiple action types?

ciampo replied:
Probably good to keep it as is, since this type describes what a `PlanAction` type is

StefanNieuwenhuis commented on packages/data-stores/src/plans/resolvers.ts:
```diff
@@ -79,13 +79,13 @@ function calculateDiscounts( planProducts: PlanProduct[] ) {
 			name: feature.name,
 			description: feature.description,
-			type: feature.type ?? 'checkbox',
+			type: 'checkbox',
```
For future reference: If the feature's type is always a checkbox, what's the added value? Do we still have to include it?

ciampo replied:
Seems like this is not needed, but we should investigare further in a separate issue #49991

StefanNieuwenhuis commented on packages/data-stores/src/plans/types.ts:
```diff
@@ -156,8 +113,11 @@ export interface Feature {
 	type: string;
 }
 
+export type DetailsAPIFeature = Omit< Feature, 'type' >;
```
Why is `Feature` an interface & `DetailsAPIFeature` a type?

ciampo replied:
Resolving in favour of https://github.com/Automattic/wp-calypso/pull/49595#discussion_r574461555

StefanNieuwenhuis commented on packages/data-stores/src/plans/types.ts:
```diff
@@ -156,8 +113,11 @@ export interface Feature {
 	type: string;
 }
 
+export type DetailsAPIFeature = Omit< Feature, 'type' >;
```
Doesn't it make more sense to let `Feature` inherit from `DetailsAPIFeature` and extend it with `type: string;` instead of the other way around? In that case, we don't have to omit props from an object. What do you think?

ciampo replied:
> What do you think?

Yep

StefanNieuwenhuis commented on packages/data-stores/src/plans/types.ts:
```diff
@@ -50,70 +50,32 @@ export interface PlanProduct {
 	raw_price: number;
-	tagline: Record< string, unknown >;
 	currency_code: string;
+	[ key: string ]: unknown;
```
Why do we want to include "unknown" props in our types (the same goes for `APIPlanDetail` & `DetailsAPIResponse`)?

ciampo replied:
As per our conversation, I moved the "unknown" types declarations in the mock files, so that the rest of our code can rely of stricter type checking

StefanNieuwenhuis reviewed (commented):
I found no TS errors and I found no logic errors. Can you please address my other comments? Thank you so much 🙏

ciampo committed:
Better declaration of Feature and DetailsAPIFeature interfaces

ciampo committed:
Move unknown properties declarations to mock data only

StefanNieuwenhuis reviewed (approved):
As per our pair-reviewing session: Approved. LGTM 🚢
